builtin/providers/azurerm/config.go

```diff
@@ -0,0 +1,214 @@
+	client.ifaceClient = ifc
+
+	lbc := network.NewLoadBalancersClient(c.SubscriptionID)
+	lbc.Authorizer = spt
+	client.loadBalancerClient = lbc
+
+	lgc := network.NewLocalNetworkGatewaysClient(c.SubscriptionID)
+	lgc.Authorizer = spt
+	client.localNetConnClient = lgc
+
+	pipc := network.NewPublicIPAddressesClient(c.SubscriptionID)
+	pipc.Authorizer = spt
+	client.publicIPClient = pipc
+
+	sgc := network.NewSecurityGroupsClient(c.SubscriptionID)
+	sgc.Authorizer = spt
+	client.secGroupClient = sgc
+
+	src := network.NewSecurityRulesClient(c.SubscriptionID)
+	src.Authorizer = spt
+	client.secRuleClient = src
+
+	snc := network.NewSubnetsClient(c.SubscriptionID)
+	snc.Authorizer = spt
+	client.subnetClient = snc
+
+	vgcc := network.NewVirtualNetworkGatewayConnectionsClient(c.SubscriptionID)
+	vgcc.Authorizer = spt
+	client.vnetGatewayConnectionsClient = vgcc
+
+	vgc := network.NewVirtualNetworkGatewaysClient(c.SubscriptionID)
+	vgc.Authorizer = spt
+	client.vnetGatewayClient = vgc
+
+	vnc := network.NewVirtualNetworksClient(c.SubscriptionID)
+	vnc.Authorizer = spt
+	client.vnetClient = vnc
+
+	rgc := resources.NewGroupsClient(c.SubscriptionID)
+	rgc.Authorizer = spt
+	client.resourceGroupClient = rgc
+
+	tc := resources.NewTagsClient(c.SubscriptionID)
+	tc.Authorizer = spt
+	client.tagsClient = tc
+
+	jc := scheduler.NewJobsClient(c.SubscriptionID)
+	jc.Authorizer = spt
+	client.jobsClient = jc
+
+	jcc := scheduler.NewJobCollectionsClient(c.SubscriptionID)
+	jcc.Authorizer = spt
+	client.jobsCollectionsClient = jcc
+
+	ssc := storage.NewAccountsClient(c.SubscriptionID)
+	ssc.Authorizer = spt
+	client.storageServiceClient = ssc
+
+	suc := storage.NewUsageOperationsClient(c.SubscriptionID)
+	suc.Authorizer = spt
+	client.storageUsageClient = suc
+
+	return &client, nil
+}
+
+// armCredentialsProvided is a helper method which indicates whether or not the
+// credentials required for authenticating against the ARM APIs were provided.
+func (c *Config) armCredentialsProvided() bool {
+	return c.ArmCredentials != "" || c._armCredentialsProvided()
+}
+func (c *Config) _armCredentialsProvided() bool {
+	return !(c.SubscriptionID == "" || c.ClientID == "" || c.ClientSecret == "" || c.TenantID == "")
+}
+
+// readArmSettings is a helper method which; given the contents of the ARM
+// credentials file, loads all the data into the Config.
+func (c *Config) readArmSettings(contents string) error {
+	data := &armConfigData{}
+	err := json.Unmarshal([]byte(contents), data)
+
+	c.SubscriptionID = data.SubscriptionID
+	c.ClientID = data.ClientID
+	c.ClientSecret = data.ClientSecret
+	c.TenantID = data.TenantID
+
+	return err
+}
+
+// validateArmConfigFile is a helper function which verifies that
+// the provided ARM configuration file is valid.
+func validateArmConfigFile(v interface{}, _ string) (ws []string, es []error) {
+	value := v.(string)
+	if value == "" {
+		return nil, nil
+	}
+
+	pathOrContents, _, err := pathorcontents.Read(v.(string))
```

Comment:
pathorcontents.Read gives back the contents as its first return value, so the variable's name is misleading...
Also, as @jen20 pointed out [on my old PR](https://github.com/hashicorp/terraform/pull/3808), wasPath should have been checked and a proper warning returned if it was indeed an unsourced file.